pkg/network/tracer/connection/tcp_close_consumer.go

```diff
@@ -75,101 +72,47 @@ func (c *tcpCloseConsumer) Stop() {
 	c.once.Do(func() {
 		close(c.closed)
 	})
 }
 
-func (c *tcpCloseConsumer) extractConn(data []byte) {
-	ct := (*netebpf.Conn)(unsafe.Pointer(&data[0]))
-	conn := c.buffer.Next()
-	populateConnStats(conn, &ct.Tup, &ct.Conn_stats, c.ch)
-	updateTCPStats(conn, &ct.Tcp_stats)
+func (c *tcpCloseConsumer) Callback(conn *network.ConnectionStats) {
+	// sentinel record post-flush
+	if conn == nil {
+		request := <-c.flushChannel
+		close(request)
+		return
+	}
+
+	closeConsumerTelemetry.perfReceived.Inc()
+	c.callback(conn)
+	c.releaser.Put(conn)
 }
 
 func (c *tcpCloseConsumer) Start(callback func(*network.ConnectionStats)) {
 	if c == nil {
 		return
 	}
-	health := health.RegisterLiveness("network-tracer")
-
-	var (
-		then             = time.Now()
-		closedCount      uint64
-		lostSamplesCount uint64
-	)
+	c.callback = callback
+	liveHealth := health.RegisterLiveness("network-tracer")
 
 	go func() {
 		defer func() {
-			err := health.Deregister()
+			err := liveHealth.Deregister()
 			if err != nil {
 				log.Warnf("error de-registering health check: %s", err)
 			}
 		}()
 
-		dataChannel := c.eventHandler.DataChannel()
-		lostChannel := c.eventHandler.LostChannel()
 		for {
 			select {
-
 			case <-c.closed:
 				return
-			case <-health.C:
-			case batchData, ok := <-dataChannel:
-				if !ok {
-					return
-				}
-
-				l := len(batchData.Data)
-				switch {
-				case l >= netebpf.SizeofBatch:
-					batch := netebpf.ToBatch(batchData.Data)
-					c.batchManager.ExtractBatchInto(c.buffer, batch)
-				case l >= netebpf.SizeofConn:
-					c.extractConn(batchData.Data)
-				default:
-					log.Errorf("unknown type received from perf buffer, skipping. data size=%d, expecting %d or %d", len(batchData.Data), netebpf.SizeofConn, netebpf.SizeofBatch)
-					continue
-				}
-
-				closeConsumerTelemetry.perfReceived.Add(float64(c.buffer.Len()))
-				closedCount += uint64(c.buffer.Len())
-				conns := c.buffer.Connections()
-				for i := range conns {
-					callback(&conns[i])
-				}
-				c.buffer.Reset()
-				batchData.Done()
-			// lost events only occur when using perf buffers
-			case lc, ok := <-lostChannel:
-				if !ok {
-					return
-				}
-				closeConsumerTelemetry.perfLost.Add(float64(lc))
-				lostSamplesCount += lc
+			case <-liveHealth.C:
 			case request := <-c.requests:
-				oneTimeBuffer := network.NewConnectionBuffer(32, 32)
-				c.batchManager.GetPendingConns(oneTimeBuffer)
-				conns := oneTimeBuffer.Connections()
-				for i := range conns {
-					callback(&conns[i])
-				}
-				close(request)
-
-				closedCount += uint64(oneTimeBuffer.Len())
-				now := time.Now()
-				elapsed := now.Sub(then)
-				then = now
-				log.Debugf(
-					"tcp close summary: closed_count=%d elapsed=%s closed_rate=%.2f/s lost_samples_count=%d",
-					closedCount,
-					elapsed,
-					float64(closedCount)/elapsed.Seconds(),
-					lostSamplesCount,
-				)
-				closedCount = 0
-				lostSamplesCount = 0
+				c.flushChannel <- request
+				c.flusher.Flush()
```

Comment:
Is `Flush` not synchronous?

Reply:
It causes the perf/ring reader to wakeup, but it doesn't do the reading itself. That is still done by `Read/ReadInto`.